### PR TITLE
[SPARK-36201][SQL][FOLLOWUP] Schema check should check inner field too

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
When inner field have wrong schema filed name should check field name too.
image

### Why are the changes needed?
Early check early faield

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @dongjoon-hyun @HyukjinKwon @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 1]
```diff
@@ -593,8 +593,14 @@ private[sql] object ParquetSchemaConverter {
        """.stripMargin.split("\n").mkString(" ").trim)
   }
 
-  def checkFieldNames(names: Seq[String]): Unit = {
-    names.foreach(checkFieldName)
+  def checkFieldNames(schema: StructType): Unit = {
+    schema.map(field => field.name -> field.dataType)
+      .foreach {
+        case (name, struct: StructType) =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala:599]: I think you can traverse with `StructField`? e.g.)

```scala
  case field if field.dataType.isInstanceOf[StructType]  =>
    checkFieldName(name)
    checkFieldNames(struct)
  case field => checkFieldName(field.name)
```

Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: Nice suggestion, done

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@AngersZhuuuu would you mind retriggering the test (https://github.com/AngersZhuuuu/spark/actions/runs/1044234803)?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu would you mind retriggering the test (https://github.com/AngersZhuuuu/spark/actions/runs/1044234803)?

Done

**Comment by HyukjinKwon** [comment]
hm, GA test seems failing.

**Comment by AngersZhuuuu** [comment]
> hm, GA test seems failing.

Seems all my UT failed like this, can't find why..

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for pinging me, @AngersZhuuuu . BTW, you should use `[FOLLOWUP]` when you reuse the JIRA id.

**Comment by dongjoon-hyun** [comment]
Wow, the GitHub Action failures look really weird.

**Comment by dongjoon-hyun** [comment]
The error code is the following. It looks like OOM happening again.
```
 ./build/mvn: line 178:  1699 Killed                  "${MVN_BIN}" "$@"
2021-07-20T04:37:47.6486105Z ##[error]Process completed with exit code 137.
```

**Comment by dongjoon-hyun** [comment]
It seems that the master branch's Java 17 job is suffering with the same reason.

**Comment by AngersZhuuuu** [comment]
> Thank you for pinging me, @AngersZhuuuu . BTW, you should use `[FOLLOWUP]` when you reuse the JIRA id.

Hmmmm, it's not a reuse jira. I write wrong jira id to `SPARK-36021` and change to correct `SPARK-36201`

**Comment by AngersZhuuuu** [comment]
> The error code is the following. It looks like OOM happening again.
> 
> * https://github.com/AngersZhuuuu/spark/runs/3110053861?check_suite_focus=true
> 
> ```
>  ./build/mvn: line 178:  1699 Killed                  "${MVN_BIN}" "$@"
> 2021-07-20T04:37:47.6486105Z ##[error]Process completed with exit code 137.
> ```

Yea, should not be related to my pr

**Comment by AngersZhuuuu** [comment]
By the way, when this can be fixed and merge this then I can start the follow up work?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala [hunk 2]
```diff
@@ -52,8 +52,13 @@ private[sql] object OrcFileFormat {
     }
   }
 
-  def checkFieldNames(names: Seq[String]): Unit = {
-    names.foreach(checkFieldName)
+  def checkFieldNames(schema: StructType): Unit = {
+    schema.foreach {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:56]: nit:
```
schema.foreach { field =>
  checkFieldName(field.name)
  field.dataType match {
    case s: StructType => checkFieldNames(s)
    case _ =>
  }
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 3]
```diff
@@ -593,8 +593,13 @@ private[sql] object ParquetSchemaConverter {
        """.stripMargin.split("\n").mkString(" ").trim)
   }
 
-  def checkFieldNames(names: Seq[String]): Unit = {
-    names.foreach(checkFieldName)
+  def checkFieldNames(schema: StructType): Unit = {
+    schema.foreach {
+      case field if field.dataType.isInstanceOf[StructType] =>
+        checkFieldName(field.name)
+        checkFieldNames(field.dataType.asInstanceOf[StructType])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala:600]: ditto

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala:56]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: Done

**Comment by cloud-fan** [comment]
GA is quite unstable now. I'm merging it to master/3.2 since jenkins has passed

**Pull request closed without merging**